README.rst
.. contents::
====================
Build RPMs for Plone
====================
Intro
-----
This repo is used to create rpms for Plone site. RPMs are created with a buildout. 

Steps
-----
* Create a rpm.cfg buildout which used to create rpms.
* Create folder named `rpm` into your buildout repo.
* Create a build.sh file into rpm folder.
* Create a job into jenkins.

Files
-----
* build.sh (replace JOB_NAME, HOME and USER) ::

    #!/bin/bash
    set -e
    export QA_RPATHS=$[ 0x0001|0x0002 ]

    RPM_VERSION="1.0.${BUILD_NUMBER}"
    RPM_NAME=project-website
    HOME=/data
    USER=userproject

    SRC_DIR=$WORKSPACE/src
    
    BUILD_DIR=$WORKSPACE/build
    rm -rf $BUILD_DIR
    mkdir $BUILD_DIR

    # directory structure needed for rpmbuild
    RPM_ROOT_DIR=$BUILD_DIR/rpm_root
    mkdir $RPM_ROOT_DIR
    cd $RPM_ROOT_DIR
    mkdir -p BUILD RPMS SRPMS SOURCES tmp

    # prepare source before archive
    ARCHIVE_DIR=$BUILD_DIR/$RPM_NAME-$RPM_VERSION
    mkdir $ARCHIVE_DIR
    cp -r $SRC_DIR/* $ARCHIVE_DIR

    # make source archive
    tar czvf $RPM_ROOT_DIR/SOURCES/$RPM_NAME-$RPM_VERSION.tar.gz --exclude=*.spec -C $BUILD_DIR \
        $RPM_NAME-$RPM_VERSION

    rpmbuild --define "name $RPM_NAME" \
        --define "home $HOME" \
        --define "user $USER" \
        --define "version $RPM_VERSION" \
        --define="_topdir $RPM_ROOT_DIR" \
        --define="_tmppath $RPM_ROOT_DIR/tmp" \
        -bb $SRC_DIR/rpm/simple.spec

    #rm -fr $BUILD_DIR


Don't forget to add execution right ::
    
    $ chmod +x build.sh

* rpm.cfg : This file is used to construct rpm.


Jobs
----
You have to construct rpms with a Jenkins job. For creating a jenkins job, 

* Go to jenkins.
* Click on *Nouveau Job*.
* Choose a name a put it into *Nom du Job* input.
* Use the option *Copier un Job existant*. In the input *Copier à partir de*, put *Build-rpm-research*
* Change the github link, put the github link to the buildout project (with ssh and not http).



CHANGES.txt
Changelog
=========

2.3.17 (unreleased)
-------------------

- Nothing changed yet.


2.3.16 (2015-07-28)
-------------------

- Extend buildout22 fix to setuptools, step2


2.3.15 (2015-07-28)
-------------------

- Extend buildout22 fix to setuptools.


2.3.14 (2015-07-28)
-------------------

- pass zc.buildout and setuptools version when running buildout.
  to shadow versions set in dist.plone.org/release/...


2.3.13 (2015-07-28)
-------------------

- Upgrade to latest zc.buildout and setuptools.


2.3.12 (2015-07-28)
-------------------

- Support latest bootstrap-buildout.


2.3.11 (2015-01-05)
-------------------

- Avoid debuginfo.


2.3.10 (2015-01-05)
-------------------

- Fix when pil*.py are not created.


2.3.9 (2015-01-05)
------------------

- Support for gscetterbeek; include oracle;


2.3.8 (2014-12-15)
------------------

- Pin setuptools and buildout versions at bootstrap time.


2.3.7 (2014-11-27)
------------------

- Update rpm version to 3.2.x to workaround ibsa jobs renamed to ploneibsa.


2.3.6 (2014-11-17)
------------------

- Fix trailing ).


2.3.5 (2014-11-17)
------------------

- Fix workaround for zc.buildout 2.2.x bug 
  when scripts do not use base variable.


2.3.4 (2014-11-05)
------------------

- Improve debug mode.
- Workaround for zc.buildout 2.2.x bug (does not install own egg in eggs
  directory).


2.3.3 (2014-05-30)
------------------

- "master" is default value for $RPMIZER_VERSION


2.3.2 (2014-05-30)
------------------

- Pass $RPMIZER_VERSION as argument of build.sh
- Improve readability of simple.spec


2.3.1 (2014-05-23)
------------------

- Run_buildout moves files instead of copying.
- Rpm build cleans more.
- Buildout config passed as parameter to run buildouts.


2.3.0 (2014-05-16)
------------------

- Add build.sh, extracted from common buildout.


2.2.0 (2013-06-07)
------------------

- Add eggs from unifiedinstaller.


2.1.4 (2013-05-14)
------------------

- Backup files are replaced now. 


2.1.3 (2013-05-14)
------------------

- Add bin/fullbackup to zeoserver rpm.


2.1.2 (2013-02-11)
------------------

- Spec files don't copy setuptools any more.


2.1.1 (2013-02-11)
------------------

- Remove bin/pil* scripts.


2.1.0 (2013-02-07)
------------------

- Sed command for changing path into interpreter script.


2.0.1 (2012-12-27)
------------------

- Add client3 and client4 for prb.
- Backup script files are no more replace by RPM update. Puppet generate this files.


2.0.0 (2012-12-21)
------------------

- Add client3 and client4.


1.0.4 (2012-12-20)
------------------

- Update prb.spec.


1.0.3 (2012-10-04)
------------------

- Add etc/ install_dir in prb.spec.


1.0.2 (2012-10-02)
------------------

- `simple.spec`
  - Use `name` instead of `portal` as input parameter
  - Add etc/ in install_dir
  - Do not distribute develop-eggs
- Add prb spec.
- Add a client2 on rpm fabrication.


1.0.1 (2012-07-20)
------------------

- Clean code : rm  $RPM_BUILD_ROOT%{installdir}/bin/copy_ckeditor_code


1.0 (2012-07-20)
----------------

- Initial release


run_buildout.sh
#!/bin/bash
set -e

if [ $# -lt 6 ]; then
    echo "Run buildout awaits path_to_python, rpm_build_directory, install_dir, buildout root, buildout version and setuptools version passed as arguments"
fi

PATH_TO_PYTHON=$1
RPM_BUILD_DIR=$2
INSTALL_DIR=$3
BUILDOUT_CFG=$4
BUILDOUT_VERSION=$5
SETUPTOOLS_VERSION=$6
GSC_ORACLE=$7

BIN_BUILDOUT="bin/buildout versions:setuptools=$SETUPTOOLS_VERSION versions:zc.buildout=$BUILDOUT_VERSION"

"$PATH_TO_PYTHON" bootstrap-buildout.py -c "$BUILDOUT_CFG" --buildout-version="$BUILDOUT_VERSION" --setuptools-version="$SETUPTOOLS_VERSION"
$BIN_BUILDOUT -N -c "$BUILDOUT_CFG" install download
$BIN_BUILDOUT -N -c "$BUILDOUT_CFG" install install
mkdir -p "$INSTALL_DIR"
mv "$RPM_BUILD_DIR/buildout-cache/downloads" "$INSTALL_DIR"
mv "$RPM_BUILD_DIR/buildout-cache/eggs" "$INSTALL_DIR"
# workaround buildout 2.2 bug where buildout does not install its own egg
# in buildout:directory/eggs
cp -r eggs/zc.buildout* "$INSTALL_DIR/eggs"
cp -r eggs/setuptools* "$INSTALL_DIR/eggs"
if [ "$GSC_ORACLE" == "true" ]; then 
  (
  export ORACLE_HOME="$INSTALL_DIR/parts/python-oracle"
  export LD_LIBRARY_PATH="$ORACLE_HOME"
  $BIN_BUILDOUT -N -c "$BUILDOUT_CFG" buildout:directory="$INSTALL_DIR"
  )
else
  $BIN_BUILDOUT -N -c "$BUILDOUT_CFG" buildout:directory="$INSTALL_DIR"
fi


install_buildout.sh
#!/bin/bash
set -e

if [ $# -lt 3 ]; then
    echo "Install buildout awaits buildout_dir, target_dir and rpm_build_root passed as arguments"
fi

BUILDOUT_DIR=$1
TARGET_DIR=$2
RPM_BUILD_ROOT=$3

# workaround buildout 2.2 bug where buildout does not install its own egg
# in buildout:directory/eggs
fix_buildout_22() {
    set +e
    grep "base = os.path.dirname(base)" "$@"
    if [ "$?" -eq "0" ]; then
        sed -i "s:^ *'[^ ]*\(eggs/zc.buildout.*egg\)',$:  join(base, '\1'),:g" "$@"
        sed -i "s:^ *'[^ ]*\(eggs/setuptools.*egg\)',$:  join(base, '\1'),:g" "$@"
    else
        sed -i "s:^ *'[^ ]*\(eggs/zc.buildout.*egg\)',$:  '$TARGET_DIR/\1',:g" "$@"
        sed -i "s:^ *'[^ ]*\(eggs/setuptools.*egg\)',$:  '$TARGET_DIR/\1',:g" "$@"
    fi
    set -e
}

INSTALL_DIR=$RPM_BUILD_ROOT$TARGET_DIR
mkdir -p "$INSTALL_DIR/etc"
mv "$BUILDOUT_DIR/bin" "$BUILDOUT_DIR/var" "$BUILDOUT_DIR/parts" "$BUILDOUT_DIR/eggs" "$INSTALL_DIR"
for file in $INSTALL_DIR/bin/*
do
    sed -i "s:${BUILDOUT_DIR/:/\\:}:${TARGET_DIR/:/\\:}:g" "$file"
    fix_buildout_22 "$file"
done
cd "$INSTALL_DIR/"
rm -fr "$INSTALL_DIR/parts/docs"
rm -fr "$INSTALL_DIR/.git"
rm "$INSTALL_DIR/bin/instance"
rm -f "$INSTALL_DIR"/bin/pil*.py
rm "$INSTALL_DIR/bin/copy_ckeditor_code"
# bin/python_oracle is needed by GSC_ETTERBEEK to build cx_Oracle
rm -f "$INSTALL_DIR/bin/python-oracle"
rm -rf "$INSTALL_DIR/parts/instance"
rm -rf "$INSTALL_DIR/parts/lxml"
find "$INSTALL_DIR" -name "*.pyc" -delete;
find "$INSTALL_DIR" -name "*.pyo" -delete;
for file in $INSTALL_DIR/parts/zeoserver/bin/*
do
    sed -i "s:${BUILDOUT_DIR/:/\\:}:${TARGET_DIR/:/\\:}:g" "$file"
    fix_buildout_22 "$file"
done
TO_CLEAN_UP=( \
    zeoserver/etc/zeo.conf \
    client1/etc/zope.conf \
    client2/etc/zope.conf \
    client3/etc/zope.conf \
    client4/etc/zope.conf \
    client1/bin/interpreter \
    client2/bin/interpreter \
    client3/bin/interpreter \
    client4/bin/interpreter \
)
for file in "${TO_CLEAN_UP[@]}"
do
    sed -i "s:${BUILDOUT_DIR/:/\\:}:${TARGET_DIR/:/\\:}:g" "$INSTALL_DIR/parts/$file"
    fix_buildout_22 "$INSTALL_DIR/parts/$file"
done


build.sh
#!/bin/bash
set -e

usage() { echo "Usage: $0 [-d] [-g] project_id" 1>&2; exit 1; }

#default values
BUILDOUT_VERSION="2.4.0"
SETUPTOOLS_VERSION="18.0.1"
RPM_BASE_VERSION="3.3"
DEBUG="build"
GSC_ORACLE="false"

if [ $# -lt 1 ]; then
    usage
fi

while getopts ":v:d:g" o; do
    case "${o}" in
        d)
            DEBUG="debug"
            ;;
        g)
            GSC_ORACLE="true"
            ;;
        *)
            usage
            ;;
    esac
done
shift $((OPTIND-1))

PROJECT_ID=${1}
RPM_NAME=${PROJECT_ID}-website
USER=${PROJECT_ID}

RPM_VERSION="${RPM_BASE_VERSION}.${BUILD_NUMBER:-undefined}"

HOME=/data


SRC_DIR=${WORKSPACE:=$(pwd)}/plone-buildout
BUILD_DIR=$WORKSPACE/build

rm -rf "$BUILD_DIR"
mkdir "$BUILD_DIR"

# directory structure needed for rpmbuild
RPM_ROOT_DIR=$BUILD_DIR/rpm_root
mkdir "$RPM_ROOT_DIR"
cd "$RPM_ROOT_DIR"
mkdir -p BUILD RPMS SRPMS SOURCES tmp

# prepare source before archive
if [ "$DEBUG" == "build" ]; then
  # Archive is needed by %setup command in specfile %prep section
  ARCHIVE_DIR=$BUILD_DIR/$RPM_NAME-$RPM_VERSION
  mkdir "$ARCHIVE_DIR"
  cp -r "$SRC_DIR"/* "$ARCHIVE_DIR"

  # make source archive
  tar czvf "$RPM_ROOT_DIR/SOURCES/$RPM_NAME-$RPM_VERSION.tar.gz" --exclude=*.spec -C "$BUILD_DIR" "$RPM_NAME-$RPM_VERSION"
fi

# get simple.spec from Rpmizer repository
SIMPLE_SPEC=$BUILD_DIR/simple.spec
if [ "$DEBUG" == "build" ]; then
  cp "$WORKSPACE/rpmizer/simple.spec" "$SIMPLE_SPEC"
else
  cp "$WORKSPACE/simple.spec" "$SIMPLE_SPEC"
fi

RUN_BUILDOUT=$BUILD_DIR/run_buildout.sh
if [ "$DEBUG" == "build" ]; then
  cp "$WORKSPACE/rpmizer/run_buildout.sh" "$RUN_BUILDOUT"
else
  cp "$WORKSPACE/run_buildout.sh" "$RUN_BUILDOUT"
fi
chmod +x "$RUN_BUILDOUT"

INSTALL_BUILDOUT=$BUILD_DIR/install_buildout.sh
if [ "$DEBUG" == "build" ]; then
  cp "$WORKSPACE/rpmizer/install_buildout.sh" "$INSTALL_BUILDOUT"
else
  cp "$WORKSPACE/install_buildout.sh" "$INSTALL_BUILDOUT"
fi
chmod +x "$INSTALL_BUILDOUT"

case $DEBUG in
  build )
    export QA_RPATHS=$(( 0x0001|0x0002 ))
    rpmbuild --define "name $RPM_NAME" \
        --define "home $HOME" \
        --define "user $USER" \
        --define "version $RPM_VERSION" \
        --define "buildout_version $BUILDOUT_VERSION" \
        --define "setuptools_version $SETUPTOOLS_VERSION" \
        --define="_topdir $RPM_ROOT_DIR" \
        --define="_tmppath $RPM_ROOT_DIR/tmp" \
        --define="run_buildout $RUN_BUILDOUT" \
        --define="install_buildout $INSTALL_BUILDOUT" \
        --define="oracle $GSC_ORACLE" \
        -bb "$SIMPLE_SPEC"
     ;;
  debug )
     BUILDOUT_DIR=$WORKSPACE/buildout
     if [ -d "$BUILDOUT_DIR" ]; then
       rm -rf "$BUILDOUT_DIR"
     fi
     mkdir -p "$BUILDOUT_DIR"
     cd "$SRC_DIR"
     rm -rf "${SRC_DIR:?}/bin" "$SRC_DIR/parts" "$SRC_DIR/.installed.cfg"
     "$RUN_BUILDOUT"  "$(which python2.7)" "$SRC_DIR" "$BUILDOUT_DIR" rpm.cfg "$BUILDOUT_VERSION" "$SETUPTOOLS_VERSION"
     TARGET_DIR=$HOME/$USER/$RPM_NAME
     RPM_BUILD_ROOT=$WORKSPACE/buildroot
     if [ -d "$RPM_BUILD_ROOT" ]; then
       rm -rf "$RPM_BUILD_ROOT"
     fi
     mkdir -p "$RPM_BUILD_ROOT"
     "$INSTALL_BUILDOUT" "$BUILDOUT_DIR" "$TARGET_DIR" "$RPM_BUILD_ROOT"
     exit 0 ;;
  * )
     echo "invalid value for DEBUG"
     exit 1 ;;
esac
